proposals/patterns3.md

```diff
@@ -0,0 +1,253 @@
+﻿# Pattern-matching changes for C# 9.0
+
+We are considering a small handful of enhancements to pattern-matching for C# 9.0 that have natural synergy and work well to address a number of common programming problems:
+- https://github.com/dotnet/csharplang/issues/2925 Type patterns
+- https://github.com/dotnet/csharplang/issues/1350 Parenthesized patterns to enforce or emphasize precedence of the new combinators
+- https://github.com/dotnet/csharplang/issues/1350 Conjunctive `and` patterns that require both of two different patterns to match;
+- https://github.com/dotnet/csharplang/issues/1350 Disjunctive `or` patterns that require either of two different patterns to match;
+- https://github.com/dotnet/csharplang/issues/1350 Negated `not` patterns that require a given pattern *not* to match; and
+- https://github.com/dotnet/csharplang/issues/812 Relational patterns that require the input value to be less than, less than or equal to, etc a given constant.
+
+## Parenthesized Patterns
+
+Parenthesized patterns permit the programmer to put parentheses around any pattern.  This is not so useful with the existing patterns in C# 8.0, however the new pattern combinators introduce a precedence that the programmer may want to override.
+
+```antlr
+primary_pattern
+    : parenthesized_pattern
+    ;
+parenthesized_pattern
+    : '(' pattern ')'
+    ;
+```
+
+## Relational Patterns
+
+Relational patterns permit the programmer to express that an input value must satisfy a relational constraint when compared to a constant value:
+
+``` C#
+    public static LifeStage LifeStageAtAge(int age) => age switch
+    {
+        < 0 =>  LiftStage.Prenatal,
+        < 2 =>  LifeStage.Infant,
+        < 4 =>  LifeStage.Toddler,
+        < 6 =>  LifeStage.EarlyChild,
+        < 12 => LifeStage.MiddleChild,
+        < 20 => LifeStage.Adolescent,
+        < 40 => LifeStage.EarlyAdult,
+        < 65 => LifeStage.MiddleAdult,
+        _ =>    LifeStage.LateAdult,
+    };
+```
+
+We imagine supporting `<`, `<=`, `>`, and `>=` patterns on all of the built-in types that support such binary relational operators with two operands of the same type in an expression. Specifically, we support all of these relational patterns for `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, and `decimal`.
+
+```antlr
+primary_pattern
+    : relational_pattern
+    ;
+relational_pattern
+    : '<' relational_expression
+    | '<=' relational_expression
+    | '>' relational_expression
+    | '>=' relational_expression
+    ;
+```
+
+The expression is required to evaluate to a constant value.  It is an error if that constant value is `double.NaN` or `float.NaN`.  It is an error if the expression is a null constant and the relational operator is `<`, `<=`, `>`, or `>=`.
```

Comment:
Is it worth specifying what would happen if you did `< double.NegativeInfinity` and related infinite constants? We could want to consider them a compile-time error?

Reply:
There is nothing special about that to specify.  If not combined with some other pattern in an `or`, such a switch case would be naturally unreachable.